reducer/src/main/java/com/graphicsfuzz/reducer/glslreducers/EliminateInjectionMacrosVisitor.java

```diff
@@ -129,15 +165,58 @@ private void cleanUpMacros(Expr parent) {
   @Override
   public void visitScalarInitializer(ScalarInitializer scalarInitializer) {
     super.visitScalarInitializer(scalarInitializer);
-    if (MacroNames.isFuzzed(scalarInitializer.getExpr())) {
+    if (MacroNames.isIdentity(scalarInitializer.getExpr())
+        || MacroNames.isZero(scalarInitializer.getExpr())
+        || MacroNames.isOne(scalarInitializer.getExpr())
+        || MacroNames.isFalse(scalarInitializer.getExpr())
+        || MacroNames.isTrue(scalarInitializer.getExpr())) {
+      scalarInitializer.setExpr(scalarInitializer.getExpr().getChild(1));
+    } else if (MacroNames.isFuzzed(scalarInitializer.getExpr())
+        || MacroNames.isDeadByConstruction(scalarInitializer.getExpr())) {
       scalarInitializer.setExpr(scalarInitializer.getExpr().getChild(0));
     }
   }
 
   private void replaceChildWithGrandchild(Expr parent, int childIndex, int grandchildIndex) {
     assert parent.getChild(childIndex).getNumChildren() > grandchildIndex;
-    parent
-        .setChild(childIndex, new ParenExpr(parent.getChild(childIndex).getChild(grandchildIndex)));
+
+    if (suitableToRemoveParentheses(parent, parent.getChild(childIndex))) {
+      parent.setChild(childIndex, parent.getChild(childIndex).getChild(grandchildIndex));
+    } else {
+      parent.setChild(childIndex,
+          new ParenExpr(parent.getChild(childIndex).getChild(grandchildIndex)));
+    }
+  }
+
+  private boolean suitableToRemoveParentheses(IAstNode parent, Expr child) {
+
+    if (child instanceof ConstantExpr
+        || child instanceof VariableIdentifierExpr
+        || child instanceof FunctionCallExpr) {
+      // It's fine to remove parentheses in cases such as (5), (x) and (sin(a)).
+      return true;
+    }
+
+    if (!(parent instanceof Expr)) {
+      // These are outer-most parentheses; fine to remove them.
+      return true;
+    }
+    if (parent instanceof ParenExpr) {
+      // These are parentheses within parentheses; fine to remove them.
+      return true;
+    }
+    if (parent instanceof FunctionCallExpr) {
```

Comment:
As the comma operator cannot appear directly as a macro argument, you should instead assert that the child is *not* an instance of the comma operator.

That would be clearer, because having code to deal with the comma operator suggests that we might expect the operator here, when in fact we do not.

Reply:
Fixed